hive/job.go

```diff
@@ -4,45 +4,70 @@ import (
 	"encoding/json"
 	"errors"
 
+	"github.com/google/uuid"
 	"github.com/suborbital/grav/grav"
 )
 
-// Job describes a job to be done
-type Job struct {
+// JobReference is a lightweight reference to a Job
+type JobReference struct {
+	uuid    string
 	jobType string
-	data    interface{}
 	result  *Result
 }
 
+// Job describes a job to be done
+type Job struct {
+	JobReference
+	data       interface{}
+	resultData interface{}
+	resultErr  error
+}
+
 // NewJob creates a new job
 func NewJob(jobType string, data interface{}) Job {
 	j := Job{
-		jobType: jobType,
-		data:    data,
+		JobReference: JobReference{
+			uuid:    uuid.New().String(),
+			jobType: jobType,
+		},
+		data: data,
 	}
 
 	return j
 }
 
+// UUID returns the Job's UUID
+func (j *JobReference) UUID() string {
+	return j.uuid
+}
+
+// Reference returns a reference to the Job
+func (j Job) Reference() *JobReference {
+	return &j.JobReference
```

Comment:
Here you are returning pointer to Job as *JobReference, I think we could make it more straightforward and provide Jobreference embeded in job as a pointer itself. We should assert that that reference will not be available to change outside the module and write  get accesor to it.
I'll explain in PR comment section why.

Reply:
Well this is a fine line that needs to be walked... I want the Job itself to NOT be a pointer, as I want the canonical object stored in the `Storage` to be non-modifiable by any other accessor, so I chose to use a non-pointer to enforce copying the value into and out of storage.

I DO however want a pointer to the JobReference, as it gets put on the heap, and the main thing that is passed through the "pipes" is a pointer rather than a full copy of the data. This reduces the amount of memory moving around.

I'm open to some tweaks of course, but in general I'm trying to optimize memory usage in this case.

Reply:
Good morning ;)

I don't know If you understand me properly so just to summarize. I totally agree with all that decision, but as I presented in code snippet below you returning pointer to Job as JobReference. So my conceptual question is, wouldn't be more logic to return pointer to JobReference instead. `return j.JobReference`
```go
type Job struct {
	*JobReference
	...
}
```
--------------
New thoughts, and to make everything clear I'll ask you some question.

JobReference is so small is it necessary to pass it as pointer, why we couldn't return it as a value.
```go
type JobReference struct {
	uuid    string
	jobType string		jobType string
	result  *Result		result  *Result
	}
```
If you returning a pointer why would you have to create lightweight version of it to pass around if you could pass a pointer to the whole Job.
Now I am thinking about leaving Job as one struct and create interface similar to JobReference and use it in Sheduler.

Reply:
I thought about this overnight and I think you're right, I'll pass the JobReference around as a value, the memory gains won't be super significant, and it makes things simpler.